File: src/main/java/com/github/sarhatabaot/kraken/core/db/ExecuteQuery.java
package com.github.sarhatabaot.kraken.core.db;

import com.github.sarhatabaot.kraken.core.extensions.LoggerExtensionsKt;
import com.lapzupi.dev.connection.ConnectionFactory;
import org.jetbrains.annotations.NotNull;
import org.jooq.DSLContext;
import org.jooq.conf.Settings;
import org.jooq.impl.DSL;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.Connection;

/**
 * @author sarhatabaot
 */
public abstract class ExecuteQuery<T, R> {
    private final Logger logger = LoggerFactory.getLogger(ExecuteQuery.class);
    private final ConnectionFactory connectionFactory;
    private final Settings settings;

    public ExecuteQuery(final ConnectionFactory connectionFactory, final Settings settings) {
        this.connectionFactory = connectionFactory;
        this.settings = settings;
    }

    public ExecuteQuery(final ConnectionFactory connectionFactory) {
        this.connectionFactory = connectionFactory;
        this.settings = null;
    }

    public T prepareAndRunQuery() {
        try (Connection connection = connectionFactory.getConnection()) {
            DSLContext dslContext = getContext(connection);
            return onRunQuery(dslContext);
        } catch (Exception e) {
            LoggerExtensionsKt.logSevereException(logger, e);
        }
        return empty();
    }

    private @NotNull DSLContext getContext(Connection connection) {
        if (settings ==null)
            return DSL.using(connection, JooqUtil.getDialect(connectionFactory.getType()));

        return DSL.using(connection, JooqUtil.getDialect(connectionFactory.getType()), settings);
    }

    public abstract T onRunQuery(DSLContext dslContext) throws Exception;

    public abstract T getQuery(@NotNull final R result) throws Exception;

    public abstract T empty();

}

File: src/main/kotlin/com/github/sarhatabaot/kraken/core/db/ExecuteQuery.kt
package com.github.sarhatabaot.kraken.core.db

import com.github.sarhatabaot.kraken.core.db.JooqUtil.getDialect
import com.github.sarhatabaot.kraken.core.extensions.logSevereException
import com.lapzupi.dev.connection.ConnectionFactory
import org.jooq.DSLContext
import org.jooq.conf.Settings
import org.jooq.impl.DSL
import org.slf4j.LoggerFactory
import java.sql.Connection

/**
 *
 * @author sarhatabaot
 */
abstract class ExecuteQuery<T, R>(
    private var connectionFactory: ConnectionFactory,
    private var settings: Settings
) {
    private val logger = LoggerFactory.getLogger(ExecuteQuery::class.java)

    open fun prepareAndRunQuery(): T {
        try {
            connectionFactory.connection.use { connection ->
                val dslContext = getContext(connection)
                return onRunQuery(dslContext)
            }
        } catch (e: Exception) {
            logger.logSevereException(e)
        }
        return empty()
    }

    private fun getContext(connection: Connection): DSLContext {
        return DSL.using(
            connection,
            getDialect(connectionFactory.type),
            settings
        )
    }

    @Throws(Exception::class)
    abstract fun onRunQuery(dslContext: DSLContext?): T

    @Throws(Exception::class)
    abstract fun getQuery(result: R): T

    abstract fun empty(): T
}